File: python/graph/hld.py
class HLD:
    def __init__(self, tree, root=0):
        n = len(tree)
        self.parent = [-1] * n
        self.depth = [0] * n
        self.heavy = [-1] * n
        self.head = [0] * n
        self.pos = [0] * n
        size = [1] * n

        def dfs(u):
            best = 0
            for v in tree[u]:
                if v == self.parent[u]:
                    continue
                self.parent[v] = u
                self.depth[v] = self.depth[u] + 1
                dfs(v)
                size[u] += size[v]
                if size[v] > best:
                    best = size[v]
                    self.heavy[u] = v

        dfs(root)
        cur = 0

        def decompose(u, h):
            nonlocal cur
            self.head[u] = h
            self.pos[u] = cur
            cur += 1
            if self.heavy[u] != -1:
                decompose(self.heavy[u], h)
                for v in tree[u]:
                    if v != self.parent[u] and v != self.heavy[u]:
                        decompose(v, v)

        decompose(root, root)

    def path_segments(self, a, b):
        ans = []
        while self.head[a] != self.head[b]:
            if self.depth[self.head[a]] < self.depth[self.head[b]]:
                a, b = b, a
            ans.append((self.pos[self.head[a]], self.pos[a] + 1))
            a = self.parent[self.head[a]]
        if self.depth[a] > self.depth[b]:
            a, b = b, a
        ans.append((self.pos[a], self.pos[b] + 1))
        return ans

File: cpp/graph/hld.cpp
#include <bits/stdc++.h>
using namespace std;
struct HLD {
  int n, t = 0;
  vector<vector<int>> g;
  vector<int> par, dep, heavy, head, pos, sz;
  HLD(int n = 0) {
    init(n);
  }
  void init(int N) {
    n = N;
    g.assign(n, {});
    par.assign(n, -1);
    dep.assign(n, 0);
    heavy.assign(n, -1);
    head = pos = vector<int>(n);
    sz.assign(n, 1);
  }
  void add_edge(int u, int v) {
    g[u].push_back(v);
    g[v].push_back(u);
  }
  int dfs(int u, int p) {
    par[u] = p;
    sz[u] = 1;
    int best = 0;
    for (int v : g[u])
      if (v != p) {
        dep[v] = dep[u] + 1;
        int s = dfs(v, u);
        sz[u] += s;
        if (s > best)
          best = s, heavy[u] = v;
      }
    return sz[u];
  }
  void decomp(int u, int h) {
    head[u] = h;
    pos[u] = t++;
    if (heavy[u] != -1)
      decomp(heavy[u], h);
    for (int v : g[u])
      if (v != par[u] && v != heavy[u])
        decomp(v, v);
  }
  void build(int r = 0) {
    dfs(r, -1);
    decomp(r, r);
  }
  template <class F> void path(int u, int v, F op, bool edge = 0) {
    while (head[u] != head[v]) {
      if (dep[head[u]] < dep[head[v]])
        swap(u, v);
      op(pos[head[u]], pos[u]);
      u = par[head[u]];
    }
    if (dep[u] > dep[v])
      swap(u, v);
    if (pos[u] + edge <= pos[v])
      op(pos[u] + edge, pos[v]);
  }
};
